components/ws-daemon/pkg/controller/controller.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/content"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/internal/session"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/iws"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/quota"
+	workspacev1 "github.com/gitpod-io/gitpod/ws-manager/api/crd/v1"
+
+	"golang.org/x/xerrors"
+	"google.golang.org/protobuf/proto"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type WorkspaceControllerOpts struct {
+	NodeName         string
+	ContentConfig    content.Config
+	UIDMapperConfig  iws.UidmapperConfig
+	ContainerRuntime container.Runtime
+	CGroupMountPoint string
+}
+
+type WorkspaceController struct {
+	client.Client
+
+	NodeName string
+
+	opts  *WorkspaceControllerOpts
+	store *session.Store
+}
+
+func NewWorkspaceController(c client.Client, opts WorkspaceControllerOpts) (*WorkspaceController, error) {
+	xfs, err := quota.NewXFS(opts.ContentConfig.WorkingArea)
+	if err != nil {
+		return nil, err
+	}
+	store, err := session.NewStore(context.Background(), opts.ContentConfig.WorkingArea, content.WorkspaceLifecycleHooks(
+		opts.ContentConfig,
+		func(instanceID string) bool { return true },
+		&iws.Uidmapper{Config: opts.UIDMapperConfig, Runtime: opts.ContainerRuntime},
+		xfs,
+		opts.CGroupMountPoint,
+	))
+	if err != nil {
+		return nil, err
+	}
+
+	return &WorkspaceController{
+		Client:   c,
+		NodeName: opts.NodeName,
+		opts:     &opts,
+		store:    store,
+	}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (wsc *WorkspaceController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&workspacev1.Workspace{}).
+		Complete(wsc)
+}
+
+func (wsc *WorkspaceController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	var workspace workspacev1.Workspace
+	if err := wsc.Get(ctx, req.NamespacedName, &workspace); err != nil {
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// not our workspace
+	if workspace.Status.Runtime == nil || workspace.Status.Runtime.NodeName != wsc.NodeName {
+		return ctrl.Result{}, nil
+	}
+
+	log.WithValues("workspaceID", workspace.Name, "runtime", workspace.Status.Runtime).Info("seen workspace we have to manage")
+
+	if workspace.Status.Phase == workspacev1.WorkspacePhaseCreating ||
+		workspace.Status.Phase == workspacev1.WorkspacePhaseInitializing ||
+		workspace.Status.Phase == workspacev1.WorkspacePhaseRunning {
+
+		if c := wsk8s.GetCondition(workspace.Status.Conditions, string(workspacev1.WorkspaceConditionContentReady)); c == nil {
+			failure, err := wsc.initWorkspaceContent(ctx, &workspace)
```

Comment:
Could the `initWorkspaceContent` function reentrant?

I ask because is there a case that the workspace is running, but the `workspace.Status.Conditions != WorkspaceConditionContentReady` because the ws-manager overwrites the `workspace.Status.Conditions`? So, the workspace content will be reinitialized?